### DISCUSSION TITLE
How come Alpine beats Vanilla JS?!

### DISCUSSION THREAD

**Discussion by delete-all** [alpinejs/alpine#3891] [2. Show and tell]
Hey all,

I have recently converted my website with basic dark-theme/navigation functionality to Alpine CSP module. The CSP build comes with around 3000 lines of code. So I thought, maybe I can simplify it with vanilla JS. I have achieved same functionality with 100 lines of code. HOWEVER, 3000 lines of Alpine code loads faster than my 100 lines of Vanilla JS code!!!

RESPECT! @calebporzio and all the contributors!

I <3 Alpine

Thanks,
Kaan

My 100 lines of vanialla JS code for your information:

`const html = document.documentElement;
    const headerLogoSource = document.getElementById('headerLogoSource');
    const headerLogoImage = document.getElementById('headerLogoImage');
    const headerToggleButton = document.getElementById('headerToggleButton');
    const stickyContainer = document.getElementById('stickyContainer');
    const stickyTemplate = document.getElementById('stickyTemplate');
    const stickyBody = document.getElementById('stickyBody');
    const SCROLL_THRESHOLD = 700;
    let dark = getThemeFromLocalStorage();

    // Initial theme setup
    updateThemeClass(dark);
    updateLogoPath(dark, headerLogoImage, headerLogoSource);
    updateTogglePath(dark, headerToggleButton);

    function appendStickyTemplate() {
        if (!stickyBody) {
            const clonedContent = document.importNode(stickyTemplate.content, true);
            stickyContainer.appendChild(clonedContent);
            // Select elements inside the cloned content after it's appended
            const stickyLogoSource = document.getElementById('stickyLogoSource');
            const stickyLogoImage = document.getElementById('stickyLogoImage');
            const stickyToggleButton = document.getElementById('stickyToggleButton');
            // Initial theme setup for sticky section
            updateLogoPath(dark, stickyLogoImage, stickyLogoSource);
            updateTogglePath(dark, stickyToggleButton);
        }
    }
    // Add scroll event listener
    window.addEventListener('scroll', function () {
        appendStickyTemplate();
        const isScrolledDown = window.pageYOffset >= SCROLL_THRESHOLD;
        toggleStickyBody(isScrolledDown);
    });

    // New function to toggle the visibility class
    function toggleStickyBody(isScrolledDown) {
        const stickyBody = document.getElementById('stickyBody');

        if (isScrolledDown) {
            stickyBody.classList.add('visible');
        } else {
            stickyBody.classList.remove('visible');
        }
    }

    // Dark theme toggle
    function handleThemeToggle() {
        dark = !dark;
        setThemeToLocalStorage(dark);
        updateThemeClass(dark);
        updateLogoPath(dark, headerLogoImage, headerLogoSource);
        updateLogoPath(dark, stickyLogoImage, stickyLogoSource);
        updateTogglePath(dark, headerToggleButton);
        updateTogglePath(dark, stickyToggleButton);
    }
    
    window.toggleTheme = handleThemeToggle;

    // Function to update the theme class on the html element
    function updateThemeClass(darkMode) {
        html.classList.toggle('theme-dark', darkMode);
    }

    // Function to update the logo path
    function updateLogoPath(darkMode, logoImage, logoSource) {
        const logoPath = getIconPath('logo', darkMode);
        logoImage.src = logoPath;
        logoSource.srcset = `${logoPath} 1x`;  // Update with appropriate sizes for different resolutions
    }

    function updateTogglePath(darkMode, toggleButton) {
        toggleButton.querySelector('path').setAttribute('d', getIconPath('toggle', darkMode));
    }

    // Function to get the theme from local storage
    function getThemeFromLocalStorage() {
        if (window.localStorage.getItem('dark')) {
            return JSON.parse(window.localStorage.getItem('dark'));
        }
        return (
            !!window.matchMedia &&
            window.matchMedia('(prefers-color-scheme: dark)').matches
        );
    }

    // Function to set the theme to local storage
    function setThemeToLocalStorage(value) {
        window.localStorage.setItem('dark', value);
    }

    // Function to get the logo path
    function getIconPath(type, darkMode) {
        if (type === 'logo') {
            return darkMode
                ? 'turncircles-logo-horizontal-white.svg'
                : 'turncircles-logo-horizontal.svg';
        } else if (type === 'toggle') {
            return darkMode
                ? 'M10 2a1 1 0 011 1v1a1 1 0 11-2 0V3a1 1 0 011-1zm4 8a4 4 0 11-8 0 4 4 0 018 0zm-.464 4.95l.707.707a1 1 0 001.414-1.414l-.707-.707a1 1 0 00-1.414 1.414zm2.12-10.607a1 1 0 010 1.414l-.706.707a1 1 0 11-1.414-1.414l.707-.707a1 1 0 011.414 0zM17 11a1 1 0 100-2h-1a1 1 0 100 2h1zm-7 4a1 1 0 011 1v1a1 1 0 11-2 0v-1a1 1 0 011-1zM5.05 6.464A1 1 0 106.465 5.05l-.708-.707a1 1 0 00-1.414 1.414l.707.707zm1.414 8.486l-.707.707a1 1 0 01-1.414-1.414l.707-.707a1 1 0 011.414 1.414zM4 11a1 1 0 100-2H3a1 1 0 000 2h1z'
                : 'M17.293 13.293A8 8 0 016.707 2.707a8.001 8.001 0 1010.586 10.586z';
        }
    }`

**Comment by ekwoka** [comment]
What is the metric by which you determine it runs faster?

**Comment by delete-all** [comment]
I've compared LCP. I've used Lighthouse and Performance Insights tools of Google Chrome and x-checked the results with GTMetrix. While GTMetrix measured slightly faster LCP with Vanilla, Lighthouse and Performance Insights measured slightly faster for Alpine. 

Here you can see the GTMetrix report for Alpine: [https://gtmetrix.com/reports/turncircles.com/rZLOQWkp/](url)  
Here you can see the GTMetrix report for Vanilla: [https://gtmetrix.com/reports/turncircles.com/Ut58JhyV/](url)

Pretty sure there is potential for improving LCP with the Vanilla code but I've concluded Alpine is great in my case; with my server and cloudflare setup...

**Reply by ekwoka** [reply]
GTMatrix is pretty much next to worthless for any kind of actual benchmarking.

But also, neither of those is actually measure the JS initialization time. You can see the alpine takes a lot longer to actually run. The download is very similar, since at max speeds, it's the RTT latency that is the bottleneck.

Alpine is good, and it can be really easy to use, but it is just NOT going to be faster than bespoke code your specific needs.

**Reply by delete-all** [reply]
You are right. No doubt that the 100lines-vanilla is lighter than Alpine. Though, given all the circumstances of my overall setup, I can conclude "there is no difference".

About the RTT, Cloudflare wants to be paid :)

What would you suggest as an alternative to GTMetrix?
